Feature/chainkit upload

cryptowizard0 opened the pull request.

cryptowizard0 committed:
feat: add initial chainkit module with redis-backed queue

- Introduces the core Chainkit struct and logic,
- Including Redis-backed upload and pending transaction management
- Interfaces for upload/download operations
- Adds comprehensive unit and benchmark tests for both the Chainkit and its Redis DB helpers.
- Adds an optgoar adapter for Arweave operations with its own test suite. This is the initial commit for the Chainkit module.

cryptowizard0 committed:
Update optgoar_test.go

cryptowizard0 committed:
​refactor(db): use unexported method names for Redis operations​

- Refactored Redis operation methods in db.goand related test files to use lowercase (unexported) method names.
- This change improves internal consistency by ensuring that these helper methods are not exposed outside the package.
- Updated all internal calls to use the new unexported method

cryptowizard0 committed:
feat(chainkit): add GraphQL support for fetching parent transaction IDs​

cryptowizard0 committed:
refactor(chainkit): update Upload and DownloadByTxid logic​

cryptowizard0 committed:
Refactor: uploading logic

cryptowizard0 committed:
feat: use cron to exec jobs

cryptowizard0 committed:
Refactor: download logic and add batch download support

Refactored the download logic to remove the parentTxID parameter and introduced new Download and Downloads methods for single and batch transaction downloads. Updated interfaces and tests accordingly. Moved GraphQL-related code to optgoar package and improved error handling and scheduler checks.

cryptowizard0 committed:
feat: Track and filter uploaded transaction IDs

Introduces Redis set storage for uploaded transaction IDs, adds methods to check and record uploaded txids, and updates upload logic to skip already uploaded txids. Includes tests for the new functionality and ensures uploaded txids are recorded after successful uploads.

cryptowizard0 committed:
remove unused func

cryptowizard0 committed:
Update chainkit_test.go

cryptowizard0 committed:
refactor: refactor redis DB logic and update Chainkit API

Refactors Redis key structure and DB logic in db.go, introducing new bundledIn status management and simplifying pending/uploaded transaction handling. Updates Chainkit methods to use new DB API, removes unused or legacy code, and adapts tests in db_test.go to the new logic. Removes the old chainkit_test.go file and adds a new check_transaction_test.go for optgoar.

cryptowizard0 committed:
refactor: modify node interface and add download verification

Replaces INodeDB with INode, updating method signatures and usages throughout the codebase. Implements download and verification logic for process transactions, including nonce range queries and assignment/message validation. Adds schema errors for process and assignment validation, removes mock_operator, and renames OptGoar GraphQL method for clarity.

cryptowizard0 committed:
feat: Improve logging and upload handling in chainkit

Enhanced debug logging throughout the chainkit package for better traceability. Refactored upload logic to generate item binaries before uploading and improved transaction status checks. Updated test cases and removed unused code for clarity and maintainability.

cryptowizard0 committed:
feat: add GraphQL query support for downloads by nonce

Introduced GraphQL-based querying for assignment and message IDs by nonce range in chainkit, including new queryByNonce and queryBatch methods. Added supporting schema types and query template in chainkit/schema/graphql.go. Updated downloadByNonce to use the new query logic and improved logging. Enhanced optgoar test to decode and print assignment details. Removed debug print from optgoar.go. Updated go.mod to include new dependencies.

cryptowizard0 committed:
feat: add owner address resolution in DownloadByPid

Updated DownloadByPid to resolve the owner's Arweave address from the spawn transaction using goarUtils.OwnerToAddress. Added debug logging for better traceability and passed the resolved address to downloadByNonce.

cryptowizard0 committed:
fix: fixed decode - miss instance for assign

cryptowizard0 committed:
remove test

cryptowizard0 committed:
feat: add local cache checks and commit for downloads

Enhanced the downloadByNonce function to check the local database for assignments and messages before downloading, and to commit new messages and assignments to Redis after download. Added CommitMessage to the INode interface and implemented it in Node, along with wiring up Chainkit initialization and handler registration in Node's constructor.

cryptowizard0 committed:
Merge branch 'develop' into feature/chainkit

cryptowizard0 committed:
fix: assignment handler registration in node initialization

Replaces AddAssignmentHandler with AddAssignResHandler when registering the chainkit assignment handler in the New function. Ensures the correct handler is used during node initialization.

cryptowizard0 committed:
feat: Refactor Chainkit config to use structured config object

Introduces a new Config struct for Chainkit configuration, moving Redis URL, keyfile path, and operator type into a unified config object. Updates initialization logic in chainkit, node, and server packages to use the new config struct. Adds config loading via Viper and updates YAML config accordingly for improved flexibility and maintainability.

cryptowizard0 committed:
Update config.yaml

cryptowizard0 committed:
feat: Add retry logic for message and assignment retrieval in recovery

Introduced getMessageAndAssignByNonce to fetch message and assignment with up to 3 retries, attempting to download missing data if retrieval fails. Updated recoveryProcess to use this new method, improving robustness during recovery operations.

cryptowizard0 committed:
remove unused

cryptowizard0 committed:
refactor: Redis queue logic and add mutex for upload

Refactors Redis usage to use a FIFO list for pending transactions and a set for uploading transactions, replacing the previous set-based approach. Adds a mutex to prevent concurrent execution of the upload check logic. Updates related methods and tests to match the new queue-based design, and simplifies the upload state machine. Also updates go.mod to add gocron v2 and adjust indirect dependencies.

cryptowizard0 committed:
add tx comfirmed log

cryptowizard0 committed:
Remove unused verification and commit methods

Commented out or removed calls to VerifyFromProcess and AuthNode in Chainkit, and deleted CommitMessage from node and INode interface. These changes reflect a shift away from certain verification and commit logic, possibly in preparation for a new local cache implementation.

cryptowizard0 committed:
Remove redundant DownloadByPid and recovery retry logic

Deleted the Node.DownloadByPid method and the getMessageAndAssignByNonce function, simplifying recovery logic to fetch messages and assignments directly from the database without retry or download attempts. This streamlines the recovery process and removes unnecessary dependencies.

cryptowizard0 committed:
Refactor fromProcess verification and cleanup recovery

Renamed VerifyFromProcess to verifyFromProcess for unexported method consistency in handle.go. Removed unused DownloadByPid call in recoveryProcess to streamline recovery logic.

cryptowizard0 committed:
delete keyfile

cryptowizard0 committed:
Remove unused download and GraphQL logic from chainkit

Deleted chainkit/download.go and chainkit/graphql.go, and removed related logic from chainkit.go. The DownloadByPid method now returns nil, and the INode interface no longer requires AuthNode. Updated node/handle.go to use a private authNode method instead of AuthNode. These changes clean up unused code and simplify the chainkit module.

cryptowizard0 committed:
Increase MaxUploadingCount to 10000

Raised the MaxUploadingCount constant from 100 to 10000 to allow more transactions in the uploading state.
Also removed unused error and struct definitions related to bundledIn status for code cleanup.

cryptowizard0 committed:
update MaxUploadingCount to 10w

hymatrix-core commented on node/node.go:
```diff
@@ -98,10 +106,16 @@ func New(
 	}
+
+	node.chainkit = chainkit.New(node, chainkitCfg)
+	node.AddAssignResHandler(node.chainkit.AssignmentHandler)
```
chainkit should be middware

hymatrix-core commented on server/server.go:
```diff
@@ -22,10 +23,10 @@ type Server struct {
 func New(
 	bundler *goar.Bundler, redisURL, arweaveURL, hymxURL string,
-	nodeInfo *schema.Info, pay *pay.Pay,
+	nodeInfo *schema.Info, pay *pay.Pay, chainkitCfg chainkitSchema.Config,
```
New(node, pay)

hymatrix-core commented on chainkit/db.go:
```diff
@@ -0,0 +1,213 @@
+	MaxUploadingCount = 100000 // 10w
+)
+
+func (c *Chainkit) moveToUploading() (int64, error) {
```
moving to db/rdb/chainkit.go

hymatrix-core commented on chainkit/schema/graphql.go:
```diff
@@ -0,0 +1,55 @@
+	Value string `json:"value"`
+}
+
+var QueryTmp = `
```
const

hymatrix-core commented on go.mod:
```diff
@@ -9,13 +9,14 @@ require (
 	github.com/gin-gonic/gin v1.10.0
-	github.com/go-co-op/gocron/v2 v2.16.3
+	github.com/go-co-op/gocron v1.37.0
+	github.com/go-co-op/gocron/v2 v2.16.5
```
no need v1, all use v2

cryptowizard0 committed:
refactor: chainkit db logic to use IDB interface and move to package rdb

Introduces a new IDB interface in schema/types.go and moves Redis-based DB logic from chainkit/db.go to db/rdb/chainkit.go. Chainkit now depends on the IDB interface and uses db/rdb/chainkit.go for Redis operations, improving modularity and testability. Updates all usages in chainkit.go and upload.go to use the new interface. Refactors and relocates tests to db/rdb/chainkit_test.go.

cryptowizard0 committed:
Update gocron to v2 and fix scheduler usage

Migrated gocron import and usage to v2 API, updating scheduler initialization, start, and shutdown methods accordingly. Also changed QueryTmp from a var to a const in schema/graphql.go.

cryptowizard0 committed:
refactor: chainkit integration and server initialization

Moved chainkit initialization out of the Node and into main, passing it explicitly to the Server. The Server now manages the lifecycle of chainkit, starting and stopping it as needed. This decouples chainkit from Node, improving modularity and responsibility separation.

cryptowizard0 committed:
Move AddAssignResHandler call to server package

The AddAssignResHandler call was removed from main.go and is now invoked within the server's Run method if chainkit is present. This centralizes handler registration logic within the server package and adds a corresponding AddAssignResHandler method to the Server struct.

hymatrix-core commented on node/node.go:
```diff
@@ -70,7 +74,7 @@ func New(
 
 	ctx, cancel := context.WithCancel(context.Background())
 
-	return &Node{
```
no need change

cryptowizard0 committed:
Update node.go
